### PR TITLE
[Analytics] Add Track function combining errors and props

### PR DESCRIPTION
**Pull request by joshheald**
Supports #5300 

## Description
For Tracks events around card reader software update events, we need to log events with both errors and other properties. To do that, it's convenient to use a track function which accepts both as parameters. This PR does not add any direct use of the combined function, but it is well exercised.

### WooAnalytics changes
Logging events with both properties, and standard error details, required a new `track(_:properties:withError:)` function on WooAnalytics. It's worth calling out here that _all_ tracks events will now go through this function so _all the scrutiny_ is welcomed!

## Testing
Test that events are logged all the way back to Tracks.

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ServiceLocator/Analytics.swift [hunk 1]
```diff
@@ -29,6 +29,15 @@ protocol Analytics {
...
     func track(_ stat: WooAnalyticsStat, withError error: Error)
 
+    /// Track a specific event with associated properties and an associated error (that is translated to properties)
+    ///
+    /// - Parameters:
+    ///   - stat: the event name
+    ///   - properties: a collection of properties related to the event
+    ///   - error: the error to track
+    ///
+    func track(_ stat: WooAnalyticsStat, properties: [AnyHashable: Any]?, withError error: Error?)
```
Comment (jaclync) [inline WooCommerce/Classes/ServiceLocator/Analytics.swift]: nit: just a suggestion
```suggestion
    func track(_ stat: WooAnalyticsStat, properties: [AnyHashable: Any]?, error: Error?)
```

File: WooCommerce/Classes/Analytics/WooAnalytics.swift [hunk 2]
```diff
@@ -98,16 +90,52 @@ public extension WooAnalytics {
...
+        let errorProperties = errorProperties(from: error)
+
+        if let passedProperties = passedProperties {
+            properties = passedProperties.merging(errorProperties ?? [:], uniquingKeysWith: { current, _ in
+                current
+            })
+        } else {
+            properties = errorProperties
+        }
+        return properties
```
Comment (jaclync) [inline WooCommerce/Classes/Analytics/WooAnalytics.swift:128]: nit: non-blocking suggestion

```suggestion
        let passedProperties: [AnyHashable: Any] = passedProperties ?? [:]
        let errorProperties = errorProperties(from: error) ?? [:]
        return passedProperties.merging(errorProperties, uniquingKeysWith: { current, _ in
            current
        })
```

File: WooCommerce/WooCommerceTests/System/WooAnalyticsTests.swift [hunk 3]
```diff
@@ -90,6 +90,36 @@ class WooAnalyticsTests: XCTestCase {
         }
     }
 
+    /// Verifies an event with an error and properties is received by the AnalyticsProvider
+    ///
+    func testEventsWithPropertiesAndErrorReceived() {
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/System/WooAnalyticsTests.swift]: nit: we use snake case for test case names now based on [this documentation](https://github.com/woocommerce/woocommerce-ios/blob/c5d83703d6c0565fc34bf410b6b27483f2318df2/docs/naming-conventions.md#test-methods) 😄 
```suggestion
    func test_events_with_properties_and_error_include_combined_properties() {
```

File: WooCommerce/WooCommerceTests/System/WooAnalyticsTests.swift [hunk 4]
```diff
@@ -90,6 +90,36 @@ class WooAnalyticsTests: XCTestCase {
...
+        let testError = NSError(domain: Constants.testErrorDomain, code: Constants.testErrorCode, userInfo: Constants.testErrorUserInfo)
+        analytics.track(.applicationOpened, properties: Constants.testProperty1, withError: testError)
+        XCTAssertEqual(testingProvider?.receivedEvents.count, 1)
+        XCTAssertEqual(testingProvider?.receivedProperties.count, 1)
+        XCTAssertEqual(testingProvider?.receivedEvents.first, WooAnalyticsStat.applicationOpened.rawValue)
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/System/WooAnalyticsTests.swift]: nit: we generally follow Given/When/Then in test cases so that it's more clear to differentiate code for each step:

```suggestion
        // Given
        let testError = NSError(domain: Constants.testErrorDomain, code: Constants.testErrorCode, userInfo: Constants.testErrorUserInfo)
        
        // When
        analytics.track(.applicationOpened, properties: Constants.testProperty1, withError: testError)
        
        // Then
        XCTAssertEqual(testingProvider?.receivedEvents.count, 1)
        XCTAssertEqual(testingProvider?.receivedProperties.count, 1)
        XCTAssertEqual(testingProvider?.receivedEvents.first, WooAnalyticsStat.applicationOpened.rawValue)
```

File: WooCommerce/WooCommerceTests/System/WooAnalyticsTests.swift [hunk 5]
```diff
@@ -90,6 +90,36 @@ class WooAnalyticsTests: XCTestCase {
...
+
+        /// Note: iOS 12 is shuffling several dictionaries (especially when it comes to serializing [:] > URL Parameters).
+        /// For that reason, we'll proceed with a bit of a more lengthy but robust check.
+        ///
+        for (key, value) in Constants.testErrorAndPropertyReceivedProperty {
+            XCTAssertEqual(value, receivedProperty1[key])
+        }
+
+        /// Second note: the error's userInfo, as a string, is getting swizzled. We'll ensure the expected payload is there,
+        /// but the exact position isn't guarranteed!
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/System/WooAnalyticsTests.swift]: nit:
```suggestion
        /// but the exact position isn't guaranteed!
```

**Review by jaclync — APPROVED** [review]
The code looks good to me! I didn't get the time to test it today, I can test it tomorrow if you'd like another confirmation 🙂

File: WooCommerce/Classes/Analytics/WooAnalytics.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/Analytics/WooAnalytics.swift:128]: Unfortunately, nil-coalescing an empty dictionary for `passedProperties` will result in a change in behaviour, which causes [tests to fail](https://github.com/woocommerce/woocommerce-ios/blob/4fb64991809c42bdd99e5edad136062d0cb8ea8d/WooCommerce/WooCommerceTests/System/WooAnalyticsTests.swift#L28) by passing an empty dictionary where previously it would have returned nil. This [affects four events](https://github.com/woocommerce/woocommerce-ios/blob/4fb64991809c42bdd99e5edad136062d0cb8ea8d/WooCommerce/Classes/Analytics/WooAnalyticsStat.swift#L545).

I approached it the same as you have here initially, but wanted to preserve the existing behaviour of the two existing functions while I combined the code into a single function.

I think it would be good to change to this more readable approach, and since we're passing an empty dictionary around it shouldn't have any impact on what gets sent to Tracks, so I'll do that in a subsequent PR.

File: WooCommerce/Classes/ServiceLocator/Analytics.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/ServiceLocator/Analytics.swift]: Yeah, agreed. I'd included `withError` to be consistent with the existing methods but just because the ones that are already there are too wordy, doesn't mean these need to be.

File: WooCommerce/WooCommerceTests/System/WooAnalyticsTests.swift (hunk 3 above)
Comment (joshheald) [reply WooCommerce/WooCommerceTests/System/WooAnalyticsTests.swift]: Thanks. Again I was going for consistency, and again I think that was the wrong call.

**Review by joshheald — COMMENTED** [review]
Thanks for the feedback and taking the time to add suggestions @jaclync
